specs/vep-1416-vdk-troubleshooting-tools/README.md

```diff
@@ -0,0 +1,100 @@
+
+# VEP-1416: VDK Troubleshooting Tools
+
+* **Author(s):** Andon Andonov (andonova@vmware.com)
+* **Status:** draft
+
+
+  - [Summary](#summary)
+  - [Glossary](#glossary)
+  - [Motivation](#motivation)
+  - [Requirements and goals](#requirements-and-goals)
+  - [High-level design](#high-level-design)
+  - [API design](#api-design)
+  - [Detailed design](#detailed-design)
+
+
+## Summary
+
+-----
+This VEP outlines the changes that would need to be introduced to provide users and Versatile Data Kit administrators with tools to troubleshoot data jobs, which are deployed in a kubernetes cluster. These tools will be provided through a dedicated plugin, which will be configurable and extendable, so more functionality could be added in the future.
+
+The initial troubleshooting capability introduced, is the ability to do a thread dump of the python process and send it to a specific endpoint.
+
+
+## Glossary
+
+-----
+- VDK: https://github.com/vmware/versatile-data-kit/wiki/dictionary#vdk
+- Plugins: https://github.com/vmware/versatile-data-kit/wiki/dictionary#vdk-plugins
+- Data Job: https://github.com/vmware/versatile-data-kit/wiki/dictionary#data-job
+- Data Job Execution: https://github.com/vmware/versatile-data-kit/wiki/dictionary#data-job-execution
+- Data Job Deployment: https://github.com/vmware/versatile-data-kit/wiki/dictionary#data-job-deployment
+- Kubernetes: https://kubernetes.io/
+- Prometheus: https://prometheus.io/docs/introduction/overview/
+
+
+## Motivation
+
+-----
+When a data job, which is deployed in a kubernetes cluster, fails with an error, the VDK Deployment administrator can troubleshoot it either through the exposed Prometheus metrics or by looking at the job's logs. This works in most cases, but fails when there are issues with the logging service, some dependency does not propagate its logs, or when the configuration of the environment has been changed.
```

Comment:
I think another problem that is hard to troubleshoot is if the job is stuck - be it due to a bug in VDK or in the user code - the job hangs .And if it runs with limited OS permissions, it may not be possible to generate even stack trace/thread dump

Reply:
Good point, note taken. When I have a working implementation, I'll test it with different container permissions and see what could be done to address issues with stuck jobs.